clippy_lints/src/unstable_intrinsics_with_stable_wrapper.rs

```diff
@@ -0,0 +1,126 @@
+use rustc_hir::def::{DefKind, Res};
+use rustc_hir::{Expr, ExprKind, Item, ItemKind, PathSegment, QPath};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::TyCtxt;
+use rustc_resolve::rustdoc::{add_doc_fragment, attrs_to_doc_fragments};
+use rustc_session::declare_lint_pass;
+use rustc_span::def_id::DefId;
+use rustc_span::{sym, Span};
+
+use clippy_utils::diagnostics::span_lint;
+use clippy_utils::{any_parent_has_attr, def_path_def_ids, match_any_def_paths};
+
+use std::sync::OnceLock;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Detects usage of unstable intrinsics with safe wrappers.
+    ///
+    /// ### Why is this bad?
+    /// It allows to have the same features without requiring to use the `core_intrinsics`
+    /// feature.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// # #![feature(core_intrinsics)]
+    /// use std::intrinsics::add_with_overflow;
+    ///
+    /// add_with_overflow(12u32, 14);
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// 12u32.overflowing_add(14);
+    /// ```
+    #[clippy::version = "1.77.0"]
+    pub UNSTABLE_INTRINSICS_WITH_STABLE_WRAPPER,
+    suspicious,
+    "Detects usage of unstable intrinsics with safe wrappers"
+}
+
+declare_lint_pass!(UnstableIntrinsicsWithStableWrapper => [UNSTABLE_INTRINSICS_WITH_STABLE_WRAPPER]);
+
+fn get_doc(tcx: TyCtxt<'_>, fn_def_id: DefId) -> String {
+    let (fragments, _) =
+        attrs_to_doc_fragments(tcx.get_attrs_unchecked(fn_def_id).iter().map(|attr| (attr, None)), true);
+    let mut doc = String::new();
+    for fragment in &fragments {
+        add_doc_fragment(&mut doc, fragment);
+    }
+    doc
+}
+
+fn emit_if_is_unstable_intrinsic(
+    cx: &LateContext<'_>,
+    def_id: DefId,
+    expr_span: Span,
+    fn_name: &str,
+    segments: &[PathSegment<'_>],
+) {
+    static FNS: OnceLock<Vec<DefId>> = OnceLock::new();
```

Comment:
Why not a `DefIdMap`?

Reply:
I don't think having a `Vec` will be a problem here, we don't have that many intrinsics anyway.

Reply:
I guess. A benchmark is always handy, I just think it's cleaner at least and more greppable